### PR TITLE
wip: Windows containerd support

### PR DESCRIPTION
**Pull request by PatrickLang**
**Reason for Change**:
Once completed, this will install ContainerD and configure Kubernetes to use it on Windows

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:
/hold do-not-merge

### CODE REVIEW

**Comment by PatrickLang** [comment]
This is going to be open a few days or more. My goal is to get the PR open early with the changes that could break Linux or fail unit tests and let the CI system find those issues. Meanwhile, I still have a lot of work to do in the Windows setup steps that aren't covered in those tests.

**Comment by PatrickLang** [comment]
This may not even build right now - did a few hours of work on a plane and then --force-with-lease

🤠

File: parts/k8s/windowsconfigfunc.ps1 [hunk 1]
```diff
@@ -109,3 +109,37 @@ function Adjust-PageFileSize()
...
+        $path += ";"+$Directory
+        $updated = $true
+    }
+    if($updated)
+    {
+        Write-Output "Updating path, added $Directory"
+        [Environment]::SetEnvironmentVariable("Path", $path, [EnvironmentVariableTarget]::Machine)
+        $env:Path = [System.Environment]::GetEnvironmentVariable("Path","Machine") + ";" + [System.Environment]::GetEnvironmentVariable("Path","User")
+    }
+}
```
Comment (PatrickLang) [inline parts/k8s/windowsconfigfunc.ps1]: @marosset - do you think I should pull these two additions into a separate PR?

File: examples/windows/kubernetes-hybrid.containerd.json [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  "apiVersion": "vlabs",
+  "properties": {
+    "orchestratorProfile": {
+      "orchestratorType": "Kubernetes",
+      "orchestratorRelease": "1.14",
+      "kubernetesConfig": {
+        "networkPlugin": "kubenet",
+        "containerRuntime": "containerd",
+        "windowsContainerdURL": "https://acsenginedev.blob.core.windows.net/cri-containerd/windows-cri-containerd-05072019.zip",
+        "windowsSdnPluginURL": "https://acsenginedev.blob.core.windows.net/cri-containerd/windows-cni-containerd-05072019.zip"
```
Comment (PatrickLang) [inline examples/windows/kubernetes-hybrid.containerd.json]: @marosset - what do you think - should these be in `kubernetesConfig`, `agentPoolProfile` or `windowsProfile` ?

File: parts/k8s/windowskubeletfunc.ps1 [hunk 3]
```diff
@@ -611,7 +638,159 @@ catch
...
+    ""dns"" : {
+    ""Nameservers"" : [ ""<NameServers>"" ],
+    ""Search"" : [ ""<Cluster DNS Suffix or Search Path>"" ]
+    },
+    ""AdditionalArgs"" : [
+    {
+        ""Name"" : ""EndpointPolicy"", ""Value"" : { ""Type"" : ""OutBoundNAT"", ""ExceptionList"": [ ""<ClusterCIDR>"", ""<MgmtSubnet>"" ] }
+    },
+    {
+        ""Name"" : ""EndpointPolicy"", ""Value"" : { ""Type"" : ""ROUTE"", ""DestinationPrefix"": ""<ServiceCIDR>"", ""NeedEncap"" : true }
```
Comment (ksubrmnn) [inline parts/k8s/windowskubeletfunc.ps1]: This should be 
```

        ""Name"" : ""EndpointPolicy"", ""Value"" : { ""Type"" : ""SDNRoute"", ""Settings"" : { ""DestinationPrefix"": ""<ServiceCIDR>"", ""NeedEncap"" : true }}
```

File: parts/k8s/windowskubeletfunc.ps1 [hunk 4]
```diff
@@ -611,7 +638,159 @@ catch
...
+        ""GW"":""<PODGW>""
+        }]
+    },
+    ""dns"" : {
+    ""Nameservers"" : [ ""<NameServers>"" ],
+    ""Search"" : [ ""<Cluster DNS Suffix or Search Path>"" ]
+    },
+    ""AdditionalArgs"" : [
+    {
+        ""Name"" : ""EndpointPolicy"", ""Value"" : { ""Type"" : ""OutBoundNAT"", ""ExceptionList"": [ ""<ClusterCIDR>"", ""<MgmtSubnet>"" ] }
```
[the remark this answers is not in the document]
Comment (ksubrmnn) [reply parts/k8s/windowskubeletfunc.ps1]: fixed in last commit

File: parts/k8s/windowskubeletfunc.ps1 (hunk 3 above)
Comment (ksubrmnn) [reply parts/k8s/windowskubeletfunc.ps1]: fixed in last commit

**Comment by PatrickLang** [comment]
Thanks @ksubrmnn - deploying a cluster with this change now

File: parts/k8s/windowskubeletfunc.ps1 (hunk 3 above)
Comment (PatrickLang) [reply parts/k8s/windowskubeletfunc.ps1]: thanks! I'm about to rebase this so no need now, but in the future it's best to run `make generate` and include that file in the same commit when changing PS1 files

**Comment by PatrickLang** [comment]
I'm getting this error in kubelet.err.log before rebasing:

```
C:\k\kubeletstart.ps1 : Cannot index into a null array.
At line:1 char:1
+ c:\k\kubeletstart.ps1
+ ~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,kubeletstart.ps1

```

Need to look into this first before rebase

**Comment by PatrickLang** [comment]
```
PS C:\k> .\kubeletstart.ps1
Ok.

Cleaning up old HNS network found
Creating a new hns Network
WARNING: The names of some imported commands from the module 'hns' include unapproved verbs that might make them less
discoverable. To find the commands with unapproved verbs, run the Import-Module command again with the Verbose
parameter. For a list of approved verbs, type Get-Verb.
VERBOSE: Invoke-HNSRequest Method[POST] Path[/networks] Data[{
    "Subnets":  [
                    {
                        "GatewayAddress":  "10.240.0.1",
                        "AddressPrefix":  "10.244.3.0/24"
                    }
                ],
    "Name":  "l2bridge",
    "Type":  "L2Bridge"
}]
VERBOSE: Result : {"Error":"A network with this name already exists. ","ErrorCode":2151350288,"Success":false}
Invoke-HnsRequest : @{Error=A network with this name already exists. ; ErrorCode=2151350288; Success=False}
At C:\k\hns.psm1:233 char:16
+ ...      return Invoke-HnsRequest -Method POST -Type networks -Data $Json ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Invoke-HNSRequest

C:\k\kubeletstart.ps1 : Cannot index into a null array.
At line:1 char:1
+ .\kubeletstart.ps1
+ ~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,kubeletstart.ps1
```

File: parts/k8s/windowskubeletfunc.ps1 [hunk 5]
```diff
@@ -605,7 +660,161 @@ catch
...
+
+    `$configJson = ConvertFrom-Json `$jsonSampleConfig
+    `$configJson.name = `$global:NetworkMode.ToLower()
+    `$configJson.ipam.subnet=`$podCIDR
+    `$configJson.ipam.routes[0].GW = `$masterSubnetGW
+    `$configJson.dns.Nameservers[0] = `$global:KubeDnsServiceIp
+    `$configJson.dns.Search[0] = `$global:KubeDnsSearchPath
+
+
+    `$configJson.AdditionalArgs[0].Value.Settings.Exceptions[0] = `$global:KubeClusterCIDR
```
Comment (PatrickLang) [inline parts/k8s/windowskubeletfunc.ps1]: @ksubrmnn - this is what I mentioned - needed to change this to match the updated schema in `$jsonSampleConfig`

Comment (ksubrmnn) [reply parts/k8s/windowskubeletfunc.ps1]: Oh man that’s frustrating. I changed it in the set up you gave me but dropped it in the new branch. Couldn’t test with no subscription :/ 

Didn’t mean to give you the run around!

**Comment by PatrickLang** [comment]
ok, another day, another rebase + squash with a commit amend to appease _Semantic Pull Request_

**Comment by PatrickLang** [comment]
/hold cancel

Ready for review. It's not done, but it builds clusters.

next steps:

1. Get valid containerd & sdnbridge builds
1. Set up test pass
1. Add Azure-CNI

File: docs/topics/clusterdefinitions.md [hunk 6]
```diff
@@ -41,13 +43,15 @@ $ aks-engine get-versions
...
 | WindowsNodeBinariesURL           | no                        | Windows Kubernetes Node binaries can be provided in the format of Kubernetes release (example: https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.11.md#node-binaries-1). This setting allows overriding the binaries for custom builds.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
+| WindowsContainerdURL             | no (for development only) | **Experimental** - see [Windows ContainerD](features.md#windows-containerd) |
```
Comment (marosset) [inline docs/topics/clusterdefinitions.md]: I don't think we should change this now but i'm curious if you think these belong here or under the WindowsProfile long-term?

File: parts/k8s/windowskubeletfunc.ps1 [hunk 7]
```diff
@@ -605,7 +664,161 @@ catch
...
+catch
+{
+    Write-Error `$_
+}
+
+"@
+    } # end elseif using sdnbridge and containerd.
+    else
+    {
+        Write-Error "The combination of $NetworkPlugin and UseContainerD=$UseContainerD is not implemented"
```
Comment (marosset) [inline parts/k8s/windowskubeletfunc.ps1]: can you throw here so we can get an error in app insights?

File: examples/windows/kubernetes-hybrid.containerd.json [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      },
+      {
+        "name": "windowspool2",
+        "count": 2,
+        "vmSize": "Standard_D2s_v3",
+        "availabilityProfile": "AvailabilitySet",
+        "osType": "Windows"
+      }
+    ],
+		"windowsProfile": {
```
[the remark this answers is not in the document]
Comment (PatrickLang) [reply examples/windows/kubernetes-hybrid.containerd.json]: done

File: docs/topics/clusterdefinitions.md (hunk 6 above)
Comment (PatrickLang) [reply docs/topics/clusterdefinitions.md]: a bit of history - this codebase originally supported mesos and docker swarm too, so the container runtime version existed before `kubernetesConfig`. 

I do think these may should have a cluster-wide default, but with a per-agentpool override. All of them happen to be in `kubernetesConfig`

- `containerRuntime`
- runtime version (`containerdVersion` / `mobyVersion`)
- `kubeletConfig`

So getting back to an answer - I'd consider moving them from `kubernetesConfig` to an agent pool setting as a group in the future for consistency, not just a single setting

File: parts/k8s/windowskubeletfunc.ps1 (hunk 7 above)
Comment (PatrickLang) [reply parts/k8s/windowskubeletfunc.ps1]: sure

Comment (PatrickLang) [reply parts/k8s/windowskubeletfunc.ps1]: done
